diff --git a/src/Personnummer.php b/src/Personnummer.php
index 7064848..f17daf4 100644
--- a/src/Personnummer.php
+++ b/src/Personnummer.php
@@ -121,7 +121,7 @@ final class Personnummer implements PersonnummerInterface
     private static function getParts(string $ssn): array
     {
         // phpcs:ignore
-        $reg = '/^(?\'century\'\d{2}){0,1}(?\'year\'\d{2})(?\'month\'\d{2})(?\'day\'\d{2})(?\'sep\'[\+\-]?)(?\'num\'(?!000)\d{3}|[TRSUWXJKLMN]\d{2})(?\'check\'\d)$/';
+        $reg = '/^(?\'century\'\d{2}){0,1}(?\'year\'\d{2})(?\'month\'\d{2})(?\'day\'\d{2})(?\'sep\'[\+\-]?)(?\'num\'(?!000)\d{3}|[TRSUWXJKLMN]\d{2}|TF[1|2])(?\'check\'\d)$/';
         preg_match($reg, $ssn, $match);

         if (empty($match)) {
@@ -284,7 +284,11 @@ final class Personnummer implements PersonnummerInterface
         }

         $num = $parts['num'];
+        $is_tf = FALSE;
         if ($this->options['allowInterimNumber'] && $this->isInterim) {
+            if ($num === 'TF1' || $num === 'TF2') {
+                $is_tf = TRUE;
+            }
             $num = preg_replace($interimTest, '1', $num);
         }

@@ -295,7 +299,10 @@ final class Personnummer implements PersonnummerInterface
         }

         $checkStr   = $parts['year'] . $parts['month'] . $parts['day'] . $num;
-        $validCheck = self::luhn($checkStr) === (int)$parts['check'];
+        // Skip check for interim numbers of type TF1 and TF2
+        $validCheck = $is_tf
+            ? is_numeric($parts['check'])
+            : self::luhn($checkStr) === (int)$parts['check'];

         return $validDate && $validCheck;
     }
